Permissions on Metadata service

adrielulanovsky opened the pull request:
## What does this change?
This implements permission checks on the following metadata-editor service endpoints for editing metadata:
```

PUT     /metadata/:id/metadata                          controllers.EditsController.setMetadata(id: String)
POST    /metadata/:id/metadata/set-from-usage-rights    controllers.EditsController.setMetadataFromUsageRights(id: String)

```
With this change, if a request to one of those endpoints is sent by a user who is not the uploader of the image AND does not have EditMetadata permission, the request will be denied.

To get the information about the uploader of the image, a new endpoint was implemented in media-api service:
`
GET     /images/:imageId/uploadedBy                   controllers.MediaApi.uploadedBy(imageId: String)
`
This endpoint returns the uploadedBy field of the image as a string. The metadata-editor service will make use of this endpoint to check if the user trying to edit metadata is the original uploader of the image.

Also, this implements some changes in the UI. In the case of uploading an image already in the system with a diffent uploader, and no EditMetadata permissions, the /upload page reflects this by showing a warning on the Image Metadata section of the conflicting image, and by disallowing edits to the metadata of that image. Batch metadata copying should work without any errors, since the batch process has been updated to avoid copying data into images without permissions.

## How can success be measured?
- As a user with EditMetadata permissions, I'm able to edit the metadata of any image.
- As a user without EditMetadata permissions, I'm only able to edit the metadata of the images I uploaded.
- When an image without EditMetadata permissions is uploaded, the user cannot edit the metadata of it, and an appropriate warning is shown in the UI.

## Screenshots
Batch upload of images, one of which is an image with a different uploader, and with no EditMetadata permissions:
![image](https://user-images.githubusercontent.com/20479781/113930702-00f14f80-97c8-11eb-93c6-757655be782b.png)


## Who should look at this?
@guardian/digital-cms @sihil @ochiengolanga 

## Tested?
- [x] locally by committer
- [ ] locally by Guardian reviewer
- [ ] on the Guardian's TEST environment

sihil commented on media-api/app/lib/elasticsearch/ElasticSearch.scala:
```diff
@@ -62,6 +62,18 @@ class ElasticSearch(val config: MediaApiConfig, mediaApiMetrics: MediaApiMetrics
+        case _ => None
+      }
+    }
+  }
```
Lovely 👍

sihil commented on rest-lib/src/main/scala/com/gu/mediaservice/lib/auth/Authorisation.scala:
```diff
@@ -22,13 +24,36 @@ class Authorisation(provider: AuthorisationProvider, executionContext: Execution
+    }
+  }
+
+  def canUserWriteMetadata(principal: Principal, uploadedBy: String): Boolean =
+    isUploaderOrHasPermission(principal, uploadedBy, EditMetadata)
+
+  def canUserDeleteImage(principal: Principal, uploadedBy: String): Boolean =
+    isUploaderOrHasPermission(principal, uploadedBy, DeleteImage)
```
Super happy for `isUploaderOrHasPermission` to be here but I'm quite keen to avoid embedding anything specific to a particular permission in this class. Can these wrapper methods be colocated in the places they are used?

adrielulanovsky replied:
Yes, I moved them here because canUserWriteMetadata is used by both media-api and metadata-editor services, but I can move the wrappers back to media-api and use isUploaderOrHasPermission directly in metadata-editor (or maybe move them to a common class? not sure where...)

adrielulanovsky replied:
Fixed

sihil commented on metadata-editor/app/controllers/EditsController.scala:
```diff
@@ -39,12 +43,43 @@ import scala.concurrent.{ExecutionContext, Future}
+  val services: Services = new Services(config.domainRoot, ServiceHosts.guardianPrefixes, Set.empty)
+  val gridClient: GridClient = GridClient(services)(ws)
+
+  def AuthorisedToEditMetadataOrUploader(imageId: String): ActionFilter[Request] = new ActionFilter[Request] {
+    override protected def filter[A](request: Request[A]): Future[Option[Result]] = {
+      //We first check for permissions, if the user has permissions we avoid making a call to media-api service
+      val hasPermission: Boolean = authorisation.hasPermissionTo(EditMetadata)(request.user)
+      if (hasPermission) {
+        Future.successful(None)
+      } else {
+        val result = for {
+          uploadedBy <- gridClient.getUploadedBy(imageId, auth.getOnBehalfOfPrincipal(request.user))
+          isAuthorised = authorisation.canUserWriteMetadata(request.user, uploadedBy.getOrElse(""))
+          if isAuthorised
+        } yield {
+          Future.successful(None)
+        }
+        result.flatten.recover{case _ => Some(authorisation.unauthorized(EditMetadata))}
+      }
+    }
+    override protected def executionContext: ExecutionContext = ec
+  }
```
This is really nice. 💯 

Without too much work I imagine that this could be made generic and moved into the Authorisation class. Perhaps an `actionFilterForUploaderOr` method? Either the method or `Authorisation` class with need a grid client.

adrielulanovsky replied:
I didn't see the use of making this generic, as the only other permission where the uploader of the image matters is DeleteImage, and the deletion of images is done directly in media-api service, so no need to do an external call. However, I may be missing some use case for that, and I'm happy to make this generic if you prefer.

adrielulanovsky replied:
I added the refactor as a new commit, showing how it could be used for DeleteImage. However, I didn't use it for that endpoint, since that one already retrieves the whole image to check for syndication rights, so it seemed kind of wasteful to do another query to elasticSearch to get the uploader of the image.

sihil reviewed (commented):
Thanks @adrielulanovsky - aside from the specific permissions in Authorisation I'm happy to merge this (after testing). Making the ActionFilter generic would be a bonus but by no means necessary.

adrielulanovsky commented on rest-lib/src/main/scala/com/gu/mediaservice/lib/auth/Authorisation.scala:
```diff
@@ -22,13 +25,61 @@ class Authorisation(provider: AuthorisationProvider, executionContext: Execution
+          true
+        } else {
+          hasPermissionTo(permission)(principal)
+        }
```
Fixed

AWare commented:
How does this look when it fails in the UI? 
To beat a dead horse, can you get an image in your session upload queue that would fail in this way- and if so, would the user know if an update failed as part of a batch update on that page.

adrielulanovsky commented:
> How does this look when it fails in the UI?
> To beat a dead horse, can you get an image in your session upload queue that would fail in this way- and if so, would the user know if an update failed as part of a batch update on that page.

Well, the only way I can think of where this could fail in the UI is if an image is loaded when a user has permission to edit metadata (so the page is loaded with the option to edit the metadata) and then the permission is disallowed before the user sends the edit. In all other cases, I believe the UI won't even show the option to edit the metadata of the image. In the former case, there is no error handling in the frontend (should we take that scenario into account?).

I'm not sure what you mean by batch updates. Could you clarify? All the uses I've seen where metadata is edited so far are individual updates on an image, not a batch process... Am I missing something?

AWare commented:
If you upload a batch of images at the same time, you can use the 
![image](https://user-images.githubusercontent.com/2670496/112879579-3956a680-90c1-11eb-85e2-b92303c91650.png) button to copy that value to all the images you've uploaded in the current session.

adrielulanovsky commented:
> If you upload a batch of images at the same time, you can use the
> ![image](https://user-images.githubusercontent.com/2670496/112879579-3956a680-90c1-11eb-85e2-b92303c91650.png) button to copy that value to all the images you've uploaded in the current session.

Oh I see, I wasn't aware of that functionality... Anyway, I don't think it will be an issue, since if you uploaded the image then you'll be able to edit the metadata no matter what permissions you have.

sihil commented:
> Oh I see, I wasn't aware of that functionality... Anyway, I don't think it will be an issue, since if you uploaded the image then you'll be able to edit the metadata no matter what permissions you have (we are checking if you have the permission OR if you're the uploader of the image).

The corner case here is that you can 'upload' an image that already exists and can currently edit metadata for that image by virtue of the fact that we do not correctly check your permissions. @paperboyo doesn't think this is used by many people, but it's used by a non-zero number of people so this will break their workflows. A workflow that we absolutely shouldn't support, but one that none the less works.

adrielulanovsky commented:
> > Oh I see, I wasn't aware of that functionality... Anyway, I don't think it will be an issue, since if you uploaded the image then you'll be able to edit the metadata no matter what permissions you have (we are checking if you have the permission OR if you're the uploader of the image).
> 
> The corner case here is that you can 'upload' an image that already exists and can currently edit metadata for that image by virtue of the fact that we do not correctly check your permissions. @paperboyo doesn't think this is used by many people, but it's used by a non-zero number of people so this will break their workflows. A workflow that we absolutely shouldn't support, but one that none the less works.

So, should we do something about it then? Display an error message when the metadata-editor controller fails? I'm not sure what we could do for the batch process though, if some fail and some don't... Perhaps show indicators on the failed images based on the id?

AWare commented:
The batching is done by scope messaging in the upload page, so if the image detail page handled it correctly, then it should _just work_ on upload. Sorry this is such a faff.

adrielulanovsky commented:
After conversation with @AWare , it was decided that the best approach would be to intercept 403 responses in Kahuna and show an appropriate error message. I attach an image for reference:

![image](https://user-images.githubusercontent.com/20479781/113315247-769a7e80-92e3-11eb-844e-1e1616e780de.png)

paperboyo commented:
> would be to intercept 403 responses in Kahuna and show an appropriate error message

The error in the screenshot is a global one. It requires the whole page to be reloaded, so that the batch edits on Upload page stop working completely. 

IIUC, this is to signpost to the user that **at least one** of the batch of images being uploaded
- already exist in Grid and
- was uploaded by somebody else, so that current user will be unable to edit its metadata (incl. as a part of batch edit to the current upload batch)

Because of that, I think the error should not be global (all other images can be edited and batch edited). And it should not cause the user to have to reload the page.

So I think the best approach would be to reuse image-specific error message, instead of a global one. Here is a crappy screenshot to illustrate what I mean. It would be shown only for affected images (all others would appear and work as is):

![image](https://user-images.githubusercontent.com/6032869/113317850-b96e4d80-9307-11eb-883e-9b671615f0be.png)
[note that the error is displayed next to the affected images only and that the metadata section is greyed out for affected images only indicating edits are impossible)

paperboyo commented:
Thanks for taking care of the UI part of this!

adrielulanovsky commented:
> > would be to intercept 403 responses in Kahuna and show an appropriate error message
> 
> The error in the screenshot is a global one. It requires the whole page to be reloaded, so that the batch edits on Upload page stop working completely.
> 
> IIUC, this is to signpost to the user that **at least one** of the batch of images being uploaded
> 
>     * already exist in Grid and
> 
>     * was uploaded by somebody else, so that current user will be unable to edit its metadata (incl. as a part of batch edit to the current upload batch)
> 
> 
> Because of that, I think the error should not be global (all other images can be edited and batch edited). And it should not cause the user to have to reload the page.
> 
> So I think the best approach would be to reuse image-specific error message, instead of a global one. Here is a crappy screenshot to illustrate what I mean. It would be shown only for affected images (all others would appear and work as is):
> 
> ![image](https://user-images.githubusercontent.com/6032869/113317850-b96e4d80-9307-11eb-883e-9b671615f0be.png)
> [note that the error is displayed next to the affected images only and that the metadata section is greyed out for affected images only indicating edits are impossible)

I removed the 403 error interceptor, and added:
- Gray-out and disabling of the Image Metadata section when uploading existing images from other uploader and without EditMetadata permissions. It should be disabled for manual editing and for batch editing as well, so no 403s should happen.
- Warning message above that section alerting the user that the image already exists in the grid uploaded by other user and they don't have permissions to edit the metadata.

Screenshot:
![image](https://user-images.githubusercontent.com/20479781/114018475-a695d480-9843-11eb-9a21-44b2d25e000e.png)

AWare commented on kahuna/public/js/upload/jobs/required-metadata-editor.html:
```diff
@@ -65,14 +67,15 @@
                 title="Apply this credit to all"
                 type="button"
-                ng-if="ctrl.withBatch"
+                ng-if="ctrl.withBatch && ctrl.userCanEdit"
```
can we also use `ng-disabled` here?

adrielulanovsky replied:
I guess we can, but if the ng-if is false I think it won't matter if it's disabled, is that right? or are you saying we replace the ng-if with ng-disabled? in that case, I think the button will still show, but clicking it won't do anything.

AWare replied:
sorry, my mistake

AWare commented on kahuna/public/js/upload/jobs/required-metadata-editor.css:
```diff
@@ -0,0 +1,3 @@
+.job-editor__disabled {
+    opacity: 0.4;
```
Appreciate it's more faff, but can we do this by altering the colour instead?

adrielulanovsky replied:
I'm not sure I understand. Do you mean I find each individual color inside the div and make it darker?

AWare replied:
yeah, it's only that grey though as you won't have invalid metadata on things already in the grid

AWare reviewed (commented):
Might also be worth greying the image. Otherwise good stuff, and I'm happy to +1 with these small changes.

AWare reviewed (approved):
tested on TEST and am happy, all Simons concerns look dealt with 👍

AWare commented:
(build looks failed because I canceled a duplicate)
